include/aws/common/byte_order.h

```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_COMMON_BYTE_ORDER_H
+#define AWS_COMMON_BYTE_ORDER_H
+
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <stdint.h>
+
+/**
+ * Returns 1 if machine is big endian, 0 if little endian.
+ * If you compile with even -O1 optimization, this check is completely optimized out at compile
+ * time and code which calls "if (is_big_endian())" will do the right thing without branching.
+ */
+static inline int is_big_endian() {
+    const uint16_t z = 0x100;
+    return *(const uint8_t*)&z;
+}
+
+/**
+ * Swaps bytes of 64 bit integer.
+ */
+static inline uint64_t byteswap64(uint64_t x) { return
+        (x << 56) | (x >> 56) |
```

Comment:
I would expect using htonl for 32-bit chunks to be faster on x86-32 as it can at least use the 32-bit bswap instruction to avoid most of the masking.

Ideal would be to use the 64-bit bswap routines in (linux-only) `endian.h`. This would need a CMake-time probe however as they're unavailable on OS X/windows. If unavailable, we'd need to fall back to inline assembler or using 32-bit htonl/ntohl.